### PR TITLE
Add the Arcade-powered source-build wrapper.

### LINKED ISSUE #2069: [ArPow] Onboard NuGet.client to arcade-powered source-build
- [x] ArPow wrapper: [in progress](https://github.com/NuGet/NuGet.Client/pull/4105) (crummel).
- [x] Introduce Arcade to NuGet repo: [in progress](https://github.com/NuGet/NuGet.Client/pull/3945) (crummel).
- [x] build.sh changes: blocked by above (NuGet team).
- [x] CI leg: [in progress](https://github.com/NuGet/NuGet.Client/pull/4105) (crummel).
- [ ] Reconcile patches: blocked by CI (shared crummel/NuGet).

### PR DESCRIPTION
**Pull request by crummel**
This adds a source-build project in eng/source-build.  This project pulls in a recent version of Arcade that supports the ArPow targets, restores arcade tooling (so we make source-build tasks available), builds NuGet in the source-build fashion (Restore/Build/Pack XPlat), and then collects the built NuGet packages into source-build intermediate packages.
Remaining work:
- publishing the intermediate packages.
- restoring dependencies from source-build intermediate packages (this requires updates to eng/Version.Details.xml as well to make sure source-build knows what the real dependencies are).
I wanted to get this in as an intermediate step because the dev workflow for source-build does work now.

There are a couple NuGet source changes I made:
- Added a global.json in the root and renamed NuGet.*C*onfig to NuGet.*c*onfig (Arcade looks for these in the root and the paths can't be easily changed.  I can try to find a workaround if this is an issue).
- Excluded trying to import some VS targets in source-build.

## Bug

Fixes: https://github.com/dotnet/source-build/issues/2069.

Regression? No.  Last working version: N/A

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 
This adds a source-build CI leg which is effectively the test of the source-build changes.

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Comment by crummel** [comment]
@dominoFire Could I get the CI label on this please?

**Comment by crummel** [comment]
ping @nkolev92 @dominoFire for the CI label please.

File: eng/SourceBuild.props [hunk 1]
```diff
@@ -0,0 +1,25 @@
+<Project>
+
+  <PropertyGroup>
+    <GitHubRepositoryName>nuget-client</GitHubRepositoryName>
+    <SourceBuildManagedOnly>true</SourceBuildManagedOnly>
+
+    <VersionPrefix>1.0.0</VersionPrefix>
+  </PropertyGroup>
+
+  <Target Name="ApplySourceBuildPatchFiles"
```
Comment (MichaelSimons) [inline eng/SourceBuild.props]: I see there are release/5.0 patches being used today.  Are these still needed?  This target is not necessary it the patches have already been integrated.

**Comment by crummel** [comment]
@dominoFire I merged https://github.com/NuGet/NuGet.Client/pull/3945 into this one so I closed that one.  cc @dseefeld and @MichaelSimons if you'd like to review.

File: global.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (nkolev92) [inline global.json:1]: I'm not sure we'd want this in our repo root. 

It makes d2d development hard and cause build failures without further changes.

File: eng/source-build/source-build.proj [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+  <PropertyGroup>
+    <ProjectRoot>$(MSBuildThisFileDirectory)../../</ProjectRoot>
+  </PropertyGroup>
+
+  <Target Name="Build" DependsOnTargets="SourceBuildPostBuild">
+    <Message Text="Running source-build Build target" Importance="High" />
+  </Target>
+
+  <Target Name="SourceBuildPreBuild">
```
Comment (MichaelSimons) [inline eng/source-build/source-build.proj]: What's the purpose of this?  Is this an extension point just in case it is needed?

Comment (crummel) [reply eng/source-build/source-build.proj]: Yes, this is where the restore and feed setup for intermediate source-build packages will go.

Comment (MichaelSimons) [reply eng/source-build/source-build.proj]: Makes sense, thanks.

File: global.json (hunk 2 above)
Comment (crummel) [reply global.json:1]: Would an empty global.json be acceptable?  From [the docs](https://docs.microsoft.com/en-us/dotnet/core/tools/global-json) it should have no effect:
> If no global.json file is found, or global.json doesn't specify an SDK version nor an allowPrerelease value, the highest installed SDK version is used (equivalent to setting rollForward to latestMajor).

Comment (nkolev92) [reply global.json:1]: I guess that'd be fine.

What's driving the requirement for a global.json?

Comment (crummel) [reply global.json:1]: Arcade looks for this in `RepoRoot`.  It wants to install the SDK specified in the global.json, but this approach bypasses that part - unfortunately the search for global.json happens much sooner.  It also expects the `.git` directory to be in the same directory, so we can't cheat it by telling it that `RepoRoot` is actually `eng/source-build`.  I replaced the contents of global.json with a comment that I hope explains this all.

File: eng/SourceBuild.props (hunk 1 above)
Comment (crummel) [reply eng/SourceBuild.props]: Removed these, the patches have been integrated.

**Comment by crummel** [comment]
@dominoFire @nkolev92 @zivkan Is there any other changes that need to be made before you can merge this?

File: build.sh [hunk 4]
```diff
@@ -40,10 +40,10 @@ DOTNET="$(pwd)/cli/dotnet"
 $DOTNET --info
 
 # Get CLI Branches for testing
-echo "dotnet msbuild build/config.props /v:m /nologo /t:GetCliBranchForTesting"
+echo "$DOTNET msbuild build/config.props /v:m /nologo /t:GetCliBranchForTesting"
```
Comment (zivkan) [inline build.sh]: `$DOTNET` points to `$RepoRoot/cli/dotnet`, which is intended for use by two of our test projects (Dotnet.Integration.Tests and MSBuild.Integration.Tests), but we prefer production/stable tooling to compile our repo and run our other scripts.

Is there some reason I can't think of why we wouldn't want to use the system-installed `dotnet`?

**Comment by zivkan** [comment]
@crummel Out of curiosity, why is `NuGet.Config` being renamed to `NuGet.config`?  [`nuget.config` is actually the preferred style on Linux, but all 3 are read](https://github.com/NuGet/NuGet.Client/blob/b83566ec2369c4e9fd07e6f95d734dfe370a1e66/src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs#L34-L36).

Furthermore, this repo is mirrored to Azure DevOps' DevDiv instance, which has "Case enforcement" enabled, and therefore this change might break mirroring.

**Comment by crummel** [comment]
This is due to an Arcade issue.  I've reverted this for now and put in a source-build patch that does it instead - I can remove this patch when we bring in intermediate package publishing to Nuget.

**Review by MichaelSimons — APPROVED** [review]

File: build.sh (hunk 4 above)
Comment (zivkan) [reply build.sh]: By the looks of it, `eng/source-build/build.sh` doesn't use `build.sh`, so I propose reverting all changes to this file.

Comment (crummel) [reply build.sh]: Reverted this as suggested, thanks.

**Review by zivkan — APPROVED** [review]
Let's hope we can get a non-flaky CI build soon.

**Review by nkolev92 — APPROVED** [review]

**Comment by crummel** [comment]
Thanks for all your help getting this in!

**Pull request merged**